File: source/glparts/FrameBuffer.java
package glparts;

import callback.EmptyCallback;
import math.Float4;
import math.Int2;
import window.GLContext;

import static org.lwjgl.opengl.GL33.*;

public class FrameBuffer extends GLObject {
    private Texture colorMap;
    private DepthTexture depthMap;
    private int buffer;

    public FrameBuffer(GLContext context, Int2 size) {
        super(null, null, context);

        colorMap = new Texture(null, null, context, size, null);
        depthMap = new DepthTexture(context, size);
        buffer = glGenFramebuffers();

        glBindFramebuffer(GL_FRAMEBUFFER, buffer);
        glFramebufferTexture2D(GL_FRAMEBUFFER, GL_COLOR_ATTACHMENT0, GL_TEXTURE_2D, colorMap.getBuffer(), 0);
        glFramebufferTexture2D(GL_FRAMEBUFFER, GL_DEPTH_ATTACHMENT, GL_TEXTURE_2D, depthMap.getBuffer(), 0);
        glBindFramebuffer(GL_FRAMEBUFFER, 0);
    }

    @Override
    public void dispose() {
        if (buffer != 0) {
            glDeleteFramebuffers(buffer);
            colorMap.dispose();
            depthMap.dispose();

            buffer = 0;
            colorMap = null;
            depthMap = null;
        }
    }

    public void clear() {
        glBindFramebuffer(GL_FRAMEBUFFER, buffer);
        glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);
        glBindFramebuffer(GL_FRAMEBUFFER, 0);
    }

    public void use(EmptyCallback callback) {
        glBindFramebuffer(GL_FRAMEBUFFER, buffer);
        callback.method();
        glBindFramebuffer(GL_FRAMEBUFFER, 0);
    }

    public Texture getColorMap() {
        return colorMap;
    }

    public DepthTexture getDepthMap() {
        return depthMap;
    }

    public void resize(Int2 size) {
        if (size.x > 0 && size.y > 0) {
            colorMap.resize(size);
            depthMap.resize(size);
        }
    }

    public Float4 getPixel(Int2 pos) {
        return getPixels(pos, new Int2(1))[0];
    }

    public Float4[] getPixels(Int2 pos, Int2 size) {
        Float4[] result = new Float4[size.x * size.y];
        float[] data = new float[result.length * 4];

        glBindFramebuffer(GL_FRAMEBUFFER, buffer);
        glReadPixels(pos.x, colorMap.getSize().y - 1 - pos.y, size.x, size.y, GL_RGBA, GL_FLOAT, data);
        glBindFramebuffer(GL_FRAMEBUFFER, 0);

        for (int i = 0; i < result.length; i++) {
            result[i] = new Float4(
                data[i * 4],
                data[i * 4 + 1],
                data[i * 4 + 2],
                data[i * 4 + 3]
            );
        }

        return result;
    }
}

File: source/glparts/FrameBuffer.kt
package glparts

import glparts.abs.GLObject
import math.Float4
import math.Int2
import org.lwjgl.opengl.GL11
import org.lwjgl.opengl.GL30
import window.GLContext

class FrameBuffer(context: GLContext?, size: Int2) : GLObject(null, null, context) {
    private val colorMap: Texture
    private val depthMap: DepthTexture
    private var buffer: Int

    init {
        colorMap = Texture(null, null, context, size, null)
        depthMap = DepthTexture(context, size)
        buffer = GL30.glGenFramebuffers()
        GL30.glBindFramebuffer(GL30.GL_FRAMEBUFFER, buffer)
        GL30.glFramebufferTexture2D(GL30.GL_FRAMEBUFFER, GL30.GL_COLOR_ATTACHMENT0, GL11.GL_TEXTURE_2D, colorMap.buffer, 0)
        GL30.glFramebufferTexture2D(GL30.GL_FRAMEBUFFER, GL30.GL_DEPTH_ATTACHMENT, GL11.GL_TEXTURE_2D, depthMap.buffer, 0)
        GL30.glBindFramebuffer(GL30.GL_FRAMEBUFFER, 0)
    }

    override fun deallocate() {
        GL30.glDeleteFramebuffers(buffer)
        buffer = 0
    }

    override fun isAllocated(): Boolean {
        return buffer != 0
    }

    fun clear() {
        GL30.glBindFramebuffer(GL30.GL_FRAMEBUFFER, buffer)
        GL11.glClear(GL11.GL_COLOR_BUFFER_BIT or GL11.GL_DEPTH_BUFFER_BIT)
        GL30.glBindFramebuffer(GL30.GL_FRAMEBUFFER, 0)
    }

    fun use(callback: () -> Unit) {
        GL30.glBindFramebuffer(GL30.GL_FRAMEBUFFER, buffer)
        callback()
        GL30.glBindFramebuffer(GL30.GL_FRAMEBUFFER, 0)
    }

    fun getColorMap(): Texture {
        return colorMap
    }

    fun getDepthMap(): DepthTexture {
        return depthMap
    }

    fun resize(size: Int2) {
        if (size.x > 0 && size.y > 0) {
            colorMap.resize(size)
            depthMap.resize(size)
        }
    }

    fun getPixel(pos: Int2): Float4? {
        return getPixels(pos, Int2(1))[0]
    }

    fun getPixels(pos: Int2, size: Int2): Array<Float4?> {
        val result = arrayOfNulls<Float4>(size.x * size.y)
        val data = FloatArray(result.size * 4)
        GL30.glBindFramebuffer(GL30.GL_FRAMEBUFFER, buffer)
        GL11.glReadPixels(pos.x, colorMap.getSize().y - 1 - pos.y, size.x, size.y, GL11.GL_RGBA, GL11.GL_FLOAT, data)
        GL30.glBindFramebuffer(GL30.GL_FRAMEBUFFER, 0)
        for (i in result.indices) {
            result[i] = Float4(
                    data[i * 4],
                    data[i * 4 + 1],
                    data[i * 4 + 2],
                    data[i * 4 + 3]
            )
        }
        return result
    }
}